### DISCUSSION TITLE
Automatically enable/disable HDR display mode to sync with client HDR

### DISCUSSION THREAD

**Discussion by anaisbetts** [LizardByte/community#149] [Sunshine]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

Similar to Gamestream, the new HDR support in nightly works when the PC and client agree on whether HDR should be enabled, but when they do not, color is distorted. This makes using HDR difficult in an environment with both HDR and SDR clients, because the colors will always be distorted on either one device or the other one without manual intervention. 

### Expected Behavior

1. Connect to Sunshine via a SDR client to an HDR-enabled display
2. If display mode for target device is HDR, switch it to SDR
3. Also the opposite

### Additional Context

Some gross code to do this can be found at https://github.com/fandangos/AutoHDR

### Host Operating System

Windows

### Operating System Version

10.0.25281.1000

### Architecture

64 bit

### Sunshine commit or version

https://github.com/LizardByte/Sunshine/commit/bcd5188ac538469e8bacd75e071b34a0d6dd4c18

### Package

Windows - installer

### GPU Type

Nvidia

### GPU Model

NVIDIA GeForce RTX 3080

### GPU Driver/Mesa Version

527.19

**Comment by hoonlight** [comment]
https://github.com/Codectory/AutoActions

This program can temporarily solve the problem. Of course, it would be better if Sunshine were improved.

**Comment by anaisbetts** [comment]
@comienzo2093 This helps, but detecting whether the incoming client is SDR or HDR is still not possible, you would have to create custom shortcuts for Steam which seems Silly

An option would be to make Sunshine run a user-provided command line when the client and server HDR settings are mismatched (and pass the client's setting as a parameter)

**Comment by anaisbetts** [comment]
Refs #825

**Comment by anaisbetts** [comment]
Alright, here's a workaround that kind of sucks but at least will get you going:

1. Copy HDRController.dll from https://1drv.ms/u/s!AjS-TuAQPSkCj5Vhx4DOKsSUstlXMA?e=izdL1v into your `C:\Program Files\Sunshine` directory (the source for this DLL can be found at https://github.com/anaisbetts/AutoActions/tree/main/Source/HDRController if you want to build it yourself)
2. Rename "Steam Big Picture" to "Steam Big Picture - SDR"
3. Add a command preparation:

Do:
`rundll32 HDRController.dll,DisableGlobalHDRState`

Undo:
`rundll32 HDRController.dll,EnableGlobalHDRState`

Create a new entry "Steam Big Picture - HDR", except set "Do" to `rundll32 HDRController.dll,EnableGlobalHDRState` and "Undo" to be blank. Also make sure to set the actual command too!

**Edit:** This is assuming you prefer your normal monitor usage to be HDR => On, if that isn't the case adjust accordingly!

**Reply by ramenatorr** [reply]
Hey @anaisbetts, I just came across and this looks super helpful for toggling HDR on and off using Sunshine. However, for whatever reason `rundll32 ` is complaining about a missing entry for the function. I'm not well-versed in Windows DLL dev to be honest. Any ideas what could cause this? Is there a step missing?

Thanks!

![image](https://github.com/LizardByte/Sunshine/assets/50680777/90ffb73f-e3ba-4526-adb1-5112d4959195)

**Reply by gpressutto5** [reply]
Is there any way to get back the value from `GetGlobalHDRState`? rundll32 won't return anything.

**Comment by Whitetigerswt** [comment]
Posting another workaround. On Windows 11, the Xbox Game Bar (Win + G) tool enables the key bind "Win + ALT + B" for toggling HDR. It also lets you create a new binding for toggling HDR. I created a new keybind "CTRL + F13" to toggle HDR (since I don't have an F13 on my keyboard, this won't conflict with anything?) I found a command prompt script that is able to simulate key strokes as well:

`echo >script.vbs set shell = CreateObject("WScript.Shell"):shell.SendKeys "^{F13}" & script.vbs`

^ = CTRL (https://ss64.com/vb/sendkeys.html)

Finally, I made a new entry in Sunshine to toggle HDR off when I connect, and back on when I disconnect:
![image](https://user-images.githubusercontent.com/1680233/216786009-27a260a0-6ac7-427e-9a10-44b787a67718.png)

This avoids downloading any sketchy code (other than that one line .BAT file I posted to send a keystroke). You shouldn't need to download any external code that you may not trust since we're relying on the Xbox Game Bar application to handle toggling HDR, which is installed on Windows 11 by default.

**Reply by tsenggordon** [reply]
Hey can you provide more details steps how to run this setup?
Is it creating a vbs file, then run a bat something like cscript scripit.vbs?
thanks!

**Reply by Vortec4800** [reply]
Thank you for posting this, it works perfectly and doesn't need extra software.

To do this, perform the following:

1) Launch the Xbox Game bar on the host by pressing Win+G.
2) Click settings in the top, then go to Shortcts and scroll to the bottom where you'll see the Toggle HDR there. Add a new custom shortcut, I used Ctrl+F12 because my keyboard didn't have an F13 key so I couldn't press it to activate it.
3) Close everything including the Game Bar. If you can test the shortcut by pressing Ctrl+F12 or whatever you set it to, that would be helpful.
4) Make a new text file with something like Notepad and paste in the one liner in the post into that file. Save it as "hdr.bat" and put it in the Sunshine directory.
5) Test the bat file by double clicking it to run and make sure HDR toggles like it should. For whatever reason it didn't work right away for me, I had to run it as administrator and then it worked. The other thing I did to test was open a Command Prompt and pasted in that same one line in the prompt and ran it manually.
6) Create a new Application in the Sunshine web interface, I called mine Steam Big Picture SDR (it's supposed to toggle off HDR in my scenario so I get HDR on my TV when I'm not using Sunshine) but you can call it whatever makes sense to you.
7) Set up the Application as shown in the screenshot with the Command just as it's shown. Make sure the path matches where you put the file. I checked "run elevated" but that may not be strictly necessary. I know what the batch file is doing so I'm not concerned about having it checked.
8) That should be it. Launch the new app from Moonlight and it will toggle HDR as expected.

**Comment by anaisbetts** [comment]
lol I shared the code, it's not _that_ sketchy :P

**Comment by mirh** [comment]
Of note that it's not just windows that you should take care of controlling, but potentially the [drivers](https://github.com/SpecialKO/SpecialK/blob/SK_23_6_2_2/src/nvapi.cpp#L411) helper libraries too. 
Also https://github.com/roshkins/IddSampleDriver/issues/23

**Comment by Maz111111** [comment]
To send the keystrokes will only toggle the current HDR mode on the host. I used nircmd to send win+alt+b as do and undo until i found this solution via the dll-thing. 
With the keystrokes i always depended on the mode that was current when starting the stream (there could be a crash before or something, so that the host was on sdr and so the keys toggle HDR or for some unknown reasons from time to time the host didn't accept the keycode win+alt+b even if typed them manually). So with this dll-solution one can be sure to toggle hdr on or off when one need it independent of the displaysettings of the host.

**Comment by Nonary** [comment]
Transferring this over to a discussion, as this is a feature request instead of a bug.

**Comment by anaisbetts** [comment]
I kind of disagree that this is a feature request? When I connect to my computer, the colors are often blown out and wrong. That is not a "feature" to have the colors be correct, that is basic functionality that is now broken in HDR.

**Reply by ReenigneArcher** [reply]
It's not a feature today, so it's a feature request. We similarly don't match the client resolution, which is along the same lines.

**Reply by anaisbetts** [reply]
This bug is worse though because unlike client resolution, there is no setting in Moonlight to instruct the server whether to enable HDR mode, and at the end of the day, it makes a feature (HDR support) unusable. A behavior that makes a feature unusable, is the definition of a bug. Because of this, HDR _doesn't work_.

**Reply by Nonary** [reply]
Actually there is a setting moonlight passes to the host to let it know that it’s requesting an HDR stream. The issue is that when you turn on HDR on the host it changes the color on the display itself in a sense. I believe that’s why Microsoft decided to prevent users from being able to turn it on without an HDR display because it would wash out the colors like it does when you stream to a non HDR device.

I suppose it might be possible for sunshine or moonlight to convert it back, but in that case it would be a feature request.

I’m not super knowledgeable about HDR, but what from what I have discussed with one of our core devs it has to do with how windows handles HDR is the main reason why the colors wash out.


I do have a script that automates turning off HDR if it detects a client that does not support it and have plans to implement it in sunshine once I get more experienced with C++.

**Reply by Nonary** [reply]
@anaisbetts Looking over the code, Sunshine already does proper tone mapping for devices without HDR enabled. The main issue is that if the user has their SDR brightness turned on, it will wash out the image. To workaround, simply setting SDR brightness to 0 will resolve this issue.

But when doing that, SDR content may be too dark on the host display. 

So basically, it's not really a bug on our side. It's Windows applying more brightness to the SDR content, since Moonlight would be streaming in SDR it would be affected by that setting in Windows.

**Comment by Nonary** [comment]
By the way here is my HDR script, it's using your modified DLL

https://github.com/Nonary/AutoHDRSwitch/releases/tag/v1

It will automatically toggle HDR if the client does not support it.
It also remembers the last setting prior to changing, so when disconnecting the stream it will go back to previous setting.

**Reply by Maz111111** [reply]
Can you please check if I understood correctly what your script is doing.
If before installing the script my non-HDR enabled SteamDeck requested the stream, will the script recognize this and turn off HDR on every request from any client from now on?
At least that's the behavior I see. Regardless of which client requests, HDR is always turned off and turned on again after the stream.
Is this intentional?
It should be the case that the script either switches HDR on or off on the host, depending on which client is requesting and which HDR mode this client wanted last time.

**Reply by Nonary** [reply]
Before you start streaming, the script will check if your client device has HDR enabled on the Moonlight settings. And then it will apply that setting to your host computers display.

After you finish streaming and disconnect, the script will put your device's color settings back to what they were before you started streaming (which, in your case, was probably turned on).

**Reply by Maz111111** [reply]
Actually, that's exactly what needs to happen.
But when I start streaming to my Shield with HDR turned on in the moonlight settings the script turns HDR off on my host. I just tested it with God of War and Uncharted. Both have an ingame-option for toggle HDR. And this option is greyed out, when HDR is set to off on the host. 
But when I stream to the steamdeck the script turns HDR off as well. So in this case it works as intended.

**Reply by Nonary** [reply]
After you start streaming with your shield with HDR turned on, can you upload your logs so I can see what is happening? Go ahead and restart Sunshine to clear the logs first, prior to doing that so it's not as much text.

You can find the logs at %windir%/temp/sunshine.log or at the WebUI in troubleshooting tab.

**Reply by Maz111111** [reply]
Thank you for the reply.  I will do this asap.

**Reply by Maz111111** [reply]
So I booted the host PC, installed your script and then started Uncharted on my shield 
The first time it crashes and closes the stream. So I start uncharted again
I tested it with an active info window of my TV - so I can see, that at the start of the stream it is HDR. After some moments (maybe when your script comes in) it switches to sdr.
After uninstalling the script and rebooting the host the stream stay on HDR

Here are the logs. I deleted my path and put a "..." instead
...

**Reply by Nonary** [reply]
@Maz111111 You did not follow the instructions, the script requires you to set your Sunshine logging level to Debug. You currently have it set to Info, so it is unable to pull the information it needs for determining if the client asked for HDR or not.

**Reply by Maz111111** [reply]
@Nonary 
I am such a stupid ...
Not only debug but also the read permission...
Now it works
Thank you

**Reply by Nonary** [reply]
@Maz111111 Awesome! Glad to hear its working for you now :)

**Comment by Maz111111** [comment]
@Nonary @anaisbetts 
I've tested it and it works great!
I would call this a breakthrough!
Thank you so much!
Please put this little script into moonlight or sunshine or where ever it has to be.
Thank you!

**EDIT:**
I only tested, streaming to a non HDR device. This works great. By now it seems, that the script only check the last client and - depending whether this client wanted HDR or not - then always sets the host to the HDR mode this client wanted (see my reply above). So it seems that the script does not (yet) take into account whether there are different clients with different HDR requirements for one host. 
But I'm not sure if the script actually wants this, or if there is a bug in the script, or if I'm using it incorrectly.

**Comment by Maz111111** [comment]
I think that there shouldn't be an option in sunshine but in moonlight like this:
"Turn off HDR in the host's system when the stream starts and turn it back on after the stream ends (only works if the host is running windows)."
This option should then cause sunshine to carry out the corresponding tasks. In the end, in my opinion, changes in the host and client software are required.
The user should actually always know which HDR mode the host has to be in for the respective client.

**Reply by anaisbetts** [reply]
I guess I don't understand why you would want this to be an option in Moonlight - it seems extremely edge-casey to ever want:

* A game in SDR when you have an HDR client (why? Bandwidth or game bugs are the only reason I can think of)
* A game in HDR when you have an SDR client (literally no reason, will always cause color issues)

This suggests that, Sunshine should just Handle this. The only option on the client side should be "Enable / Disable HDR support", which would force an HDR-capable client to declare itself as SDR (which would cover the above edge-case Nr. 1)

**Reply by Nonary** [reply]
Also like I mentioned earlier, if you set SDR content brightness to zero, then it doesn't really matter if HDR is on or off on the client. Sunshine and Moonlight will properly convert the colors, it's the SDR brightness feature that causes the washout.

Which, if using a dummy plug, is an acceptable solution in this case as well.

**Reply by Maz111111** [reply]
@anaisbetts 
Yes, you are right.
I turn HDR off, when I stream rpcs3 because the emulator has a strange behaviour. It switches to SDR when it starts a game but when it switches back to HDR after quitting the game my display is greyed out and only the active window ist "highlighted".
This does not happen, when I turn off HDR manually before starting an emulated game.
But this seems to be a problem of rpcs3 and is really an edge-case. And because one doesn't want to change the setting before every streaming session, one will always have to handle this in the application section of sunshine for this certain game with a do and undo by using your command.
So your logic is totally correct.

**Comment by cgutman** [comment]
> Similar to Gamestream, the new HDR support in nightly works when the PC and client agree on whether HDR should be enabled, but when they do not, color is distorted. This makes using HDR difficult in an environment with both HDR and SDR clients, because the colors will always be distorted on either one device or the other one without manual intervention.

@anaisbetts I think you may have stumbled upon a latent bug in our handling of Advanced Color SDR displays. While HDR can't be streamed color-accurately to SDR clients, clients with HDR enabled should always be able to receive color-accurate SDR streams (and should adjust their display mode accordingly to the HDR state of the host). In other words, it's expected to be fine to leave HDR enabled in Moonlight all the time, regardless of the state of HDR on the host.

I have a fix for this issue in #1178

**Reply by anaisbetts** [reply]
@cgutman No, this is still an issue afaik, though the issue is the opposite of what you describe - if you have an HDR server and an SDR client, it will still be bugged; this is expected (because ofc you can't squish a larger color space into a smaller one as you say).

My point though is, what should happen is that Sunshine should switch the Server screen mode to disable HDR when it sees that the *client* is SDR-only, so that the _content_ (i.e. game) inherently renders in SDR. This means, that Sunshine will get SDR content and can send it along to the client.

(I have a HDR monitor that I prefer to leave in HDR all the time, but I have both HDR and SDR clients -  when I connect via the HDR client the server doesn't have to do anything, but when I connect via the SDR client, it should switch the mode, launch the game, then switch it back when the game exits).

**Reply by cgutman** [reply]
Correct, I didn't mean the whole issue was fixed in general. I just meant that the specific case of SDR host to HDR client was broken in cases where it shouldn't have been (one of the possible HDR-SDR mismatches that you mentioned in your post).

We'll do the HDR switching as part of the work to support changing the display resolution on the host when streaming. I hope to work on that in the coming days/weeks.

**Reply by anaisbetts** [reply]
That's awesome to hear! Thanks for all your work on this incredible project <3

**Comment by da-anda** [comment]
isn't this rather a bug in Moonlight which would have to do tonemapping if it receives a HDR signal from the server, but the client only has SDR capability? Or would tonemapping add to much additional latency? Because at the end of the day, sunshine clients (like moonlight) basically are just fancy low latency video players, and any other video player processing HDR content on a SDR screen also has to do tonemapping.

**Reply by blackgold9** [reply]
No, you want to eliminate the tonemapping at the source. If HDR is turned on on the source machine, the video is rendered differently, windows autohdr might be activated, and then the client will (Likely inaccurately) have to tonemap it back to sdr. If the server changes it's displaymode to match the client, then no mapping needs to be done and you get a better picture.

**Comment by anthqiu** [comment]
Windows 11 supports HDR quick toggle using key combination `Win` `Alt` `B`. This can be simulated using [AutoIt](https://www.autoitscript.com/site/autoit/):

```autoit
Send("#!b")
```

I have made an exe file using Aut2exe so it can run directly without AutoIt installation: [hdr.zip](https://github.com/LizardByte/Sunshine/files/13858926/hdr.zip).

**Comment by mist0706** [comment]
Has this been worked on? I'm still getting overblown colors if i don't manually disable HDR on the host computer before connecting
